# 1. Change Identity

In [42]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df1 = pd.read_excel('SQL_Pandas_Task.xlsx',sheet_name='Exam 1',header=1)

In [3]:
df2 = pd.read_excel('SQL_Pandas_Task.xlsx',sheet_name='Exam 2',header=1)

In [4]:
df3 = pd.concat([df1,df2])

In [5]:
subjects = [x for x in df3['Exam name'].unique()]*39
names = sorted([x for x in df3['Student name'].unique()]*6)

In [6]:
change_identifications = pd.DataFrame(subjects,names)

In [7]:
change_identifications = change_identifications.reset_index()

In [8]:
change_identifications.columns = ['Student name','Exam name']

In [9]:
exam_1 = []
exam_2 = []
flag = []
for x in range(len(change_identifications)):
    try:
        e1 = int(df1[(df1['Student name']==change_identifications.iloc[x]['Student name'])& (df1['Exam name']==change_identifications.iloc[x]['Exam name'])]['Exam points'])
    except:
        e1 = 'Absent'
    try:
        e2 = int(df2[(df2['Student name']==change_identifications.iloc[x]['Student name'])& (df2['Exam name']==change_identifications.iloc[x]['Exam name'])]['Exam points'])
    except:
        e2 = 'Absent'

    exam_1.append(e1)
    exam_2.append(e2)

    if e1 == e2 and e1 !='Absent':
        flag.append('No Change')
    elif e1 =='Absent' and e2 !='Absent':
        flag.append('Insert')
    elif e1 !='Absent' and e2 =='Absent':
        flag.append('Delete')
    elif e1==e2=='Absent':
        flag.append('Absent for both')
    else:
        flag.append('Updates')

In [10]:
change_identifications['Exam1 points'] = exam_1
change_identifications['Exam2 points'] = exam_2
change_identifications['Flags'] = flag

# Solution dataframe for 1st problem

In [11]:
change_identifications

Student name    Exam name Exam1 points Exam2 points            Flags
0     Akeem Chan      Biology           41           62          Updates
1     Akeem Chan    Chemistry           38       Absent           Delete
2     Akeem Chan  Mathematics           67           17          Updates
3     Akeem Chan   Philosophy           33           17          Updates
4     Akeem Chan      Physics       Absent       Absent  Absent for both
..           ...          ...          ...          ...              ...
229   Yen Bowers    Chemistry       Absent       Absent  Absent for both
230   Yen Bowers  Mathematics           40           98          Updates
231   Yen Bowers   Philosophy           36            8          Updates
232   Yen Bowers      Physics       Absent       Absent  Absent for both
233   Yen Bowers    Sociology           52           22          Updates

[234 rows x 5 columns]

# 2.

### List of students who appeared for all subject exams as part of exam 2

In [12]:
value_count = df2['Student name'].value_counts()

In [13]:
l = [value_count[x] for x in df2['Student name']]

In [14]:
df2['exam_count']  = l

In [15]:
df_all_six = df2[df2['exam_count']==6]

In [16]:
df_all_six

Student name  Exam name  Exam points  exam_count
6       Shannon Fields    Biology           38           6
7       Jennifer Keith    Biology           87           6
8            Keane Cox    Biology           27           6
9          Sage Foster    Biology           44           6
10         Ezra Hoover    Biology           29           6
..                 ...        ...          ...         ...
178       Galena Mayer  Sociology           67           6
179        Reese Black  Sociology            6           6
180       Phillip Ward  Sociology           72           6
181  Priscilla Higgins  Sociology           14           6
182      Yael Trujillo  Sociology           66           6

[120 rows x 4 columns]

# Sum total marks and percentage for each student

In [17]:
df_all_six.groupby('Student name').sum()['Exam points']

Student name
Alma Richmond        293
Burton Burke         321
Ezra Hoover          336
Galena Mayer         350
Grady Decker          98
Hayley Hull          288
Jenette Humphrey     343
Jennifer Keith       414
Keane Cox            188
Kirestin Sweet       219
Mohammad Cantrell    315
Octavia Monroe       384
Octavius Chambers    257
Phillip Ward         428
Preston Barrera      256
Priscilla Higgins    241
Reese Black          109
Sage Foster          260
Shannon Fields       411
Yael Trujillo        449
Name: Exam points, dtype: int64

In [18]:
df_all_six.groupby('Student name').sum()['Exam points']*100/600

Student name
Alma Richmond        48.833333
Burton Burke         53.500000
Ezra Hoover          56.000000
Galena Mayer         58.333333
Grady Decker         16.333333
Hayley Hull          48.000000
Jenette Humphrey     57.166667
Jennifer Keith       69.000000
Keane Cox            31.333333
Kirestin Sweet       36.500000
Mohammad Cantrell    52.500000
Octavia Monroe       64.000000
Octavius Chambers    42.833333
Phillip Ward         71.333333
Preston Barrera      42.666667
Priscilla Higgins    40.166667
Reese Black          18.166667
Sage Foster          43.333333
Shannon Fields       68.500000
Yael Trujillo        74.833333
Name: Exam points, dtype: float64

In [19]:
d = pd.DataFrame(df_all_six.groupby('Student name').sum()['Exam points']*100/600)

In [20]:
d.columns = ['percentage']

In [21]:
d['total_marks'] = df_all_six.groupby('Student name').sum()['Exam points']

In [22]:
sorted_d = d.sort_values('total_marks',ascending=False)

In [23]:
sorted_d['rank'] = range(1,21)

# Final solution for (2.1 & 2.2)

In [24]:
sorted_d

percentage  total_marks  rank
Student name                                    
Yael Trujillo       74.833333          449     1
Phillip Ward        71.333333          428     2
Jennifer Keith      69.000000          414     3
Shannon Fields      68.500000          411     4
Octavia Monroe      64.000000          384     5
Galena Mayer        58.333333          350     6
Jenette Humphrey    57.166667          343     7
Ezra Hoover         56.000000          336     8
Burton Burke        53.500000          321     9
Mohammad Cantrell   52.500000          315    10
Alma Richmond       48.833333          293    11
Hayley Hull         48.000000          288    12
Sage Foster         43.333333          260    13
Octavius Chambers   42.833333          257    14
Preston Barrera     42.666667          256    15
Priscilla Higgins   40.166667          241    16
Kirestin Sweet      36.500000          219    17
Keane Cox           31.333333          188    18
Reese Black         18.166667          109    19
Grady Decker        16.333333           98    20

# 90 percentile

In [25]:
topStudents90 = int((100-90)*len(sorted_d)/100)  # formula for 90 percentile 

In [26]:
topStudents90

2

In [27]:
d = dict()
l = []
for x in df2['Exam name'].unique():
    tab = df_all_six[df_all_six['Exam name']==x].sort_values('Exam points',ascending=False)[:topStudents90]
    d[x]=tab.iloc[topStudents90-1]['Exam points']
    d[f'Above 90 in {x}'] = tab['Student name'].to_list()
#     l.extend(tab['Student name'].to_list())

# Solution for 2.3

In [28]:
d

{'Biology': 87,
 'Above 90 in Biology': ['Phillip Ward', 'Jennifer Keith'],
 'Chemistry': 83,
 'Above 90 in Chemistry': ['Alma Richmond', 'Galena Mayer'],
 'Mathematics': 91,
 'Above 90 in Mathematics': ['Octavia Monroe', 'Sage Foster'],
 'Philosophy': 93,
 'Above 90 in Philosophy': ['Yael Trujillo', 'Phillip Ward'],
 'Physics': 97,
 'Above 90 in Physics': ['Shannon Fields', 'Phillip Ward'],
 'Sociology': 95,
 'Above 90 in Sociology': ['Hayley Hull', 'Shannon Fields']}

# 3. list of studennt who appeared atleast 4 subject exam

In [29]:
df2=df2.drop('exam_count',axis=1)

In [30]:
df4 = df3.groupby(['Student name','Exam name']).mean()

In [31]:
exam_appeared = df4.groupby('Student name').count()


In [32]:
exam_appeared.columns = ['exam_appeared']

In [33]:
df5 = pd.concat([exam_appeared,df4.groupby('Student name').mean()],axis=1)

In [34]:
df6 = df5[df5['exam_appeared']>=4]

In [35]:
df6.shape

(36, 2)

In [36]:
df7 = df6.sort_values('Exam points',ascending=False)

In [37]:
df7['rank'] = range(1,len(df7)+1)

# 3.1 Rank of students based on average score

In [38]:
df7

exam_appeared  Exam points  rank
Student name                                         
Jennifer Keith                   6    64.500000     1
Phillip Ward                     6    63.000000     2
Colin Adkins                     4    62.875000     3
Octavia Monroe                   6    62.083333     4
Rogan Edwards                    4    60.750000     5
Galena Mayer                     6    60.583333     6
Hayley Hull                      6    58.333333     7
Shannon Fields                   6    58.333333     8
Yael Trujillo                    6    58.250000     9
Portia Hamilton                  4    57.750000    10
Jenette Humphrey                 6    55.750000    11
Brooke Zimmerman                 4    55.750000    12
Demetrius Castaneda              6    54.916667    13
Ryan Pena                        4    54.750000    14
Cadman Greene                    4    54.250000    15
Darrel Wallace                   4    53.875000    16
Todd Dillard                     5    53.300000    17
Mohammad Cantrell                6    51.916667    18
Noelle Marsh                     5    51.500000    19
Preston Barrera                  6    49.916667    20
Hop Cantu                        5    49.900000    21
Victoria Long                    4    49.375000    22
Akeem Chan                       5    49.000000    23
Burton Burke                     6    48.666667    24
Ezra Hoover                      6    46.250000    25
Yen Bowers                       4    46.000000    26
Octavius Chambers                6    44.250000    27
Priscilla Higgins                6    44.000000    28
Keane Cox                        6    44.000000    29
Alma Richmond                    6    41.416667    30
Kirestin Sweet                   6    40.166667    31
Reese Black                      6    39.666667    32
Sage Foster                      6    39.000000    33
Grady Decker                     6    38.916667    34
Georgia Gomez                    5    37.400000    35
Nita Valdez                      5    35.800000    36

# 3.2 Get mean ,median and mode

# Mean

In [39]:
df7['Exam points'].mean()

51.00532407407408

# Median

In [40]:
df7['Exam points'].median()

51.70833333333333

# Mode

In [41]:
df7['Exam points'].mode()

0    44.000000
1    55.750000
2    58.333333
dtype: float64

# 1.The MEAN and MEDIAN of the Exam points are symmetrically distribution that why the value are same.



# 2.The MODE is the value that appeared most often in a set of data values, the set of 44;55.75;58.33 is frequent.